File: quicksort.js
function qs(array, startIdx = 0, length = array.length) {
  if (length <= 1) {
    return array;
  }

  let pivot = array[startIdx];
  let barrierIdx = startIdx;

  for (let i = startIdx + 1; i <= length + startIdx; i++) {
    if (pivot > array[i]) {
      barrierIdx++;
      let swap1 = array[barrierIdx];
      array[barrierIdx] = array[i];
      array[i] = swap1;
    }
  }

  array[startIdx] = array[barrierIdx];
  array[barrierIdx] = pivot;


  let left = barrierIdx - startIdx;
  let right = length - (left + 1);
  qs(array, startIdx, left); //left
  qs(array, barrierIdx + 1, right); //right

  return array;

}


// console.log(qs([2, 6, 5, 8, 1, 67, 0, 3, 23]));

function quickSort(arr) {
  if (arr.length < 2) {
    return arr;
  }

  let pivot = arr[0];
  let left = arr.slice(1, arr.length).filter((el) => el < pivot);
  let right = arr.slice(1, arr.length).filter((el) => el > pivot);
  
  let sorted = quickSort(left).concat(pivot).concat(quickSort(right));
  
  return sorted;
  
  
}

console.log(quickSort([2, 6, 5, 8, 1, 67, 0, 3, 23]));

function inPlaceQS(arr, startIdx = 0, length = arr.length) {
  if (length < 2) {
    return arr; 
  }

  let pivot = arr[startIdx];
  let barrier = startIdx;
  for (let i = startIdx + 1; i <= length + startIdx; i++) {
    if (arr[i] < pivot) {
      barrier++;
      let temp = arr[barrier];
      arr[barrier] = arr[i];
      arr[i] = temp;
    }
  }
  
  arr[startIdx] = arr[barrier];
  arr[barrier] = pivot;
  
  let leftLength = barrier - startIdx;
  let rightLength = length - (leftLength + 1);
  inPlaceQS(arr, 0, leftLength);
  inPlaceQS(arr, barrier + 1, rightLength);
 
  return arr;
}

console.log(inPlaceQS([2, 6, 5, 8, 1, 67, 0, 3, 23]));

File: quick_sort.rb
#recursion, requires base case, and something that peels the layers
# left + pivot + right 
# pivot = first 
# return left + pivot + right 

#time complexity worst case is O(n2), this is if you already have an sorted array such as [1, 2, 3, 4, 5]. You will go through each number then n deep
#best case scenario is O(nlogn) if each sub array is constantly even when splitted
#space complexity is O(logn) since the size gets smaller and smalelr each time 
def quick_sort(arr)
  
  if arr.length <= 1
    return arr
  end 

  pivot = arr[0]
  left = []
  right = []

  #seperate into left and right 
  arr[1..-1].each do |num|
    if num < pivot
      left.push(num)
    else
      right.push(num)
    end  
  end 

  result = quick_sort(left) + [pivot] + quick_sort(right)

  return result 

end 


p quick_sort([3, 2, 1])